datajoint/external.py

```diff
@@ -0,0 +1,132 @@
+
+import boto3
+from botocore.exceptions import ClientError
+
+from . import config
+from . import DataJointError
+
+logger = logging.getLogger(__name__)
+
+
+def bucket(aws_access_key_id=None, aws_secret_access_key=None, reset=False):
+    """
+    Returns a boto3 AWS session object to be shared by multiple modules.
+    If the connection is not yet established or reset=True, a new
+    connection is set up. If connection information is not provided,
+    it is taken from config which takes the information from
+    dj_local_conf.json. If the password is not specified in that file
+    datajoint prompts for the password.
+
+    :param aws_access_key_id: AWS Access Key ID
+    :param aws_secret_access_key: AWS Secret Key
+    :param reset: whether the connection should be reset or not
+    """
+    if not hasattr(bucket, 'bucket') or reset:
+        aws_access_key_id = aws_access_key_id \
+            if aws_access_key_id is not None \
+            else config['external.aws_access_key_id']
+
+        aws_secret_access_key = aws_secret_access_key \
+            if aws_secret_access_key is not None \
+            else config['external.aws_secret_access_key']
+
+        if aws_access_key_id is None:  # pragma: no cover
+            aws_access_key_id = input("Please enter AWS Access Key ID: ")
+
+        if aws_secret_access_key is None:  # pragma: no cover
+            aws_secret_access_key = getpass(
+                "Please enter AWS Secret Access Key: "
+            )
+
+        bucket.bucket = Bucket(aws_access_key_id, aws_secret_access_key)
+    return bucket.bucket
+
+
+class Bucket:
+    """
+    A dj.Bucket object manages a connection to an AWS S3 Bucket.
+
+    Currently, basic CRUD operations are supported; of note permissions and
+    object versioning are not currently supported.
+
+    Most of the parameters below should be set in the local configuration file.
+
+    :param aws_access_key_id: AWS Access Key ID
+    :param aws_secret_access_key: AWS Secret Key
+    """
+
+    def __init__(self, aws_access_key_id, aws_secret_access_key):
+        self._session = boto3.Session(
+            aws_access_key_id=aws_access_key_id,
+            aws_secret_access_key=aws_secret_access_key
+        )
+        self._s3 = None
+        try:
+            self._bucket = config['external.location'].split("s3://")[1]
+        except (AttributeError, IndexError, KeyError) as e:
+            raise DataJointError('external.location not properly configured: '
+                                 + str(config['external.location'])) from None
+
+    def connect(self):
+        if self._s3 is None:
+            self._s3 = self._session.resource('s3')
+
+    def stat(self, rpath=None):
+        """
+        check if a file exists in the bucket
+        """
+        try:
+            self.connect()
+            self._s3.Object(self._bucket, rpath).load()
+        except ClientError as e:
+            if e.response['Error']['Code'] == "404":
```

Comment:
Had some thoughts about this when implementing - to some extent it depends on 

a) consistency with API in other extfile stuff
b) notion of default processing vs exceptions

this call is asking 'if a file exists' - 
False indicates it does not; an exception indicates the program was not able to successfully check.

Essentially this logic flips the S3 API to more closely mimic something like stat(2) so that things like:

```
if not mybucket.stat('/remote/path'):
     mybucket.put('/local/path', '/remote/path')
```
work rather than needing to do:
```
try:
   mybucket.stat('/remote/path')
except aws.NoS3FileErrorExceptionGizmo:
   mybucket.put('/local/path','/remote/path')
finally:
   raise DataJointError('your s3 doesnt work')
```

Reply:
misunderstanding corrected, as is the code :)